ENG-1183 Update to SDK dev.05 and use CLI (#10)
nextmv-io/tour: building-models/searching.md

you to define the guardrails to generate new stores from existing ones. This
creates a search tree that the solver uses to find the best operationally valid
store. A store is operationally valid if all decisions have been made and they
are valid.

We are going to search for all permutations of the positive integer numbers
that go up to a specific number, e.g.:

 ```txt
1 -> [
  [1]
]
2 -> [
  [1,2],
  [2,1]
]
3 -> [
  [1,2,3]
  [1,3,2]
  [2,1,3]
  [2,3,1]
  [3,1,2]
  [3,2,1]  
]
 ```

You can see that as we increase the number, this search becomes non-trivial.
Starting with an integer input `n`, define the root store and a domain of the
unused integers:

```go
root := store.New()
unused := store.NewDomain(root, model.NewRange(1, n))
```

We are going to use a slice to store the permutations as we search for them.

```go
permutation := store.NewSlice[int](root)
```

Our store is operationally valid if we have found _all_ permutations, i.e.: the
`unused` domain is empty.

```go
root = root.Validate(unused.Empty)
```

For a simple output format, we can visualize the `permutation` slice.

```go
root = root.Format(func(s store.Store) any { return permutation.Slice(s) })
```

> From an existing parent store, we must define to rules to generate children

Reviewer: ```suggestion
From an existing parent store, we must define the rules to generate child
```

stores. We can do so through an `Eager` or `Lazy` generator. In the following

> code snippet, we lazily generate new children stores by getting the values that

Reviewer: ```suggestion
code snippet, we lazily generate new child stores by getting the values that
```

[3,2,1]
```

> Use the Nextmv cli to look at complete examples of more advanced search

Reviewer 2: capitalize CLI

problems:

```bash
$ nextmv sdk init -t knapsack
Successfully generated the knapsack template.
$ nextmv sdk init -t sudoku
Successfully generated the sudoku template.
$ nextmv sdk init -t shift-scheduling
Successfully generated the shift-scheduling template.
```

## Exercises

> * Run the turorial for different values of `n`, such as 4.

Reviewer 3: ```suggestion
* Run the tutorial for different values of `n`, such as 4.
```
typo

Reviewer: I think `tuROARial` would make a great name for tutorials that we are very proud of 😆
